client/lib/protect-form/index.tsx

```diff
@@ -13,7 +13,7 @@ import { Subtract } from 'utility-types';
  */
 const debug = debugModule( 'calypso:protect-form' );
 
-type ComponentMarkedWithFormChanges = Component;
+type ComponentMarkedWithFormChanges = Component | string;
```

Comment:
I'm totally unsure whether this is a good idea. Afaics consumers would have to make sure that the `id` passed to the hook is unique. Should we rather incorporate this into the custom hook?

Reply:
I think it's an unneccessary complication. All we need is that the `useProtectForm` hook instance, and also the `ProtectFormGuard` and `protectForm()` component instances, have some unique ID. And that can be simply an unique instance of `{}`.

```js
type FormId = {};

const formsChanged = new Set<FormId>();

function useProtectForm() {
  const formId = useRef<FormId>( {} );

  const markChanged = useCallback( () => formsChanged.add( formId.current ), [] );
  const markSaved = useCallback( () => formsChanged.delete( formId.current ), [] );

  return { markChanged, markSaved };
}
```

The class component forms of this can either be reimplemented with the hook, or can store an unique `{}` as a ref, too, and call `markChanged( this.idRef.current )` instead of `markChanged( this )`.

Reply:
Makes total sense to me, but eslint warns about using `{}` as a type ([`@typescript-eslint/ban-types`](https://github.com/typescript-eslint/typescript-eslint/blob/v4.22.0/packages/eslint-plugin/docs/rules/ban-types.md))

```
Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
- If you want a type meaning "empty object", you probably want `Record<string, never>`
```

In this case it seems the first option is what we want.

Reply:
Added https://github.com/Automattic/wp-calypso/pull/52054/commits/aea31ca80780e9a7a65c22c3f6b2a723575f9c13 which implements your suggestions but I can sure use some feedback on the TypeScript changes.

Reply:
Even the `Record<string, never>` variant would be fine, as the value is always an empty object. Only there are many unique instances of it.

One way to get rid of this problem is to use `type FormId = []`. Empty array does the same job, and doesn't offend the linter.